src/Our.Umbraco.Ditto/ComponentModel/Processors/UmbracoXPathAttribute.cs

```diff
@@ -0,0 +1,98 @@
+using System.Linq;
+using Umbraco.Core;
+using Umbraco.Core.Models;
+using Umbraco.Web;
+
+namespace Our.Umbraco.Ditto
+{
+    /// <summary>
+    /// A processor that queries the Umbraco content cache using an XPath expression.
+    /// </summary>
+    public class UmbracoXPathAttribute : DittoProcessorAttribute
+    {
+        /// <summary>
+        /// A dictionary lookup of XPath variable names, along with corresponding functions.
+        /// </summary>
+        protected Dictionary<string, Func<IPublishedContent, string>> Lookup { get; set; }
+
+        /// <summary>
+        /// The XPath expression used to query the Umbraco content cache.
+        /// </summary>
+        public string XPath { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="UmbracoXPathAttribute"/> class.
+        /// </summary>
+        public UmbracoXPathAttribute()
+        {
+            // This lookup attempts to simplify how Umbraco core handles the XPath syntax parsing
+            // ref: https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Core/Xml/UmbracoXPathPathSyntaxParser.cs
+
+            Lookup = new Dictionary<string, Func<IPublishedContent, string>>()
+            {
+                { "$current", x => string.Format("id({0})", x.Id) },
+                { "$parent", x => string.Format("id({0})", x.Parent.Id) },
+                { "$site", x => string.Format("id({0})", x.Path.ToDelimitedList().ElementAtOrDefault(1) ?? "-1") },
+                { "$root", x => "/root" },
+            };
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="UmbracoXPathAttribute"/> class.
+        /// </summary>
+        /// <param name="xpath">The XPath expression used to query the Umbraco content cache.</param>
+        public UmbracoXPathAttribute(string xpath)
+            : this()
+        {
+            XPath = xpath;
+        }
+
+        /// <summary>
+        /// Processes the value.
+        /// </summary>
+        /// <returns>
+        /// Returns the objects that match the XPath expression from the Umbraco content cache.
+        /// </returns>
+        public override object ProcessValue()
+        {
+            if (string.IsNullOrWhiteSpace(XPath))
+                return Value;
+
+            var content = GetPublishedContent(Value);
+
+            if ((content == null || content.Id == 0) && UmbracoContext.Current.PageId.HasValue)
+                content = UmbracoContext.Current.ContentCache.GetById(UmbracoContext.Current.PageId.Value);
+
+            var xparam = Lookup.FirstOrDefault(x => XPath.StartsWith(x.Key));
+
+            var xpath = xparam.Key != null
+                ? XPath.Replace(xparam.Key, xparam.Value(content))
+                : XPath;
+
+            return UmbracoContext
+                .Current
+                .ContentCache
+                .GetByXPath(xpath)
+                .OrderBy(x => x.Level)
+                .ThenBy(x => x.SortOrder);
```

Comment:
My only concern (it's a big one) with this processor is that it allows you to too easily add navigation properties without protecting you from the overheads of doing so. 

While on the surface to a consuming developer it might look like we are simply doing the same as the pickers, we are most definitely not. parsing `$site`, `$root` etc involves document traversal which should be a method call rather than a property since they are so expensive. Sure you can mark it as lazy but who will?